### PR TITLE
CValidationInterface: ValidationInterfaceUnregistering, called when being unregistered

### PR DESCRIPTION
**Pull request by luke-jr**
Part of https://github.com/bitcoin-core/gui/pull/444 (GUI Network Watch tool)

### CODE REVIEW

**Review by vincenzopalazzo — COMMENTED** [review]
Concept ACK

**Comment by jonatack** [comment]
Concept ACK, have been looking at this code and plan to review.

**Comment by pk-b2** [comment]
Concept ACK 8eed7dd722177c2e36645d66e2c8254f9037a0e3

File: src/validationinterface.cpp [hunk 1]
```diff
@@ -139,6 +139,8 @@ void UnregisterSharedValidationInterface(std::shared_ptr<CValidationInterface> c
 
 void UnregisterValidationInterface(CValidationInterface* callbacks)
 {
+    callbacks->ValidationInterfaceUnregistering();
```
Comment (pk-b2) [inline src/validationinterface.cpp:142]: Is it worth to be more specific in the naming whether the unregistering will happen or has already happened?

Like renaming the function to `::ValidationInterfaceWillUnregister`?

This makes it clearer that the callback hasn't been removed yet.

Comparing to introducing a callback which gets called after the fact of the callback being removed `::ValidationInterfaceDidUnregister'

File: src/validationinterface.cpp [hunk 2]
```diff
@@ -149,6 +151,9 @@ void UnregisterAllValidationInterfaces()
     if (!g_signals.m_internals) {
         return;
     }
+
+    g_signals.m_internals->Iterate([&](CValidationInterface& callbacks) { callbacks.ValidationInterfaceUnregistering(); });
```
Comment (pk-b2) [inline src/validationinterface.cpp:155]: 👍 Follows the same approach as in `::UnregisterValidationInterface` of being called before the callback gets removed from the list

File: src/validationinterface.h [hunk 3]
```diff
@@ -173,6 +173,13 @@ class CValidationInterface {
      * Notifies listeners that a block which builds directly on our current tip
      * has been received and connected to the headers tree, though not validated yet */
     virtual void NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock>& block) {};
+    /**
+     * Notifies the validation interface that it is being unregistered
+     */
+    virtual void ValidationInterfaceUnregistering() {};
```
Comment (pk-b2) [inline src/validationinterface.h:179]: You don't need it in https://github.com/bitcoin-core/gui/pull/444 but is it worth to consider to add `::ValidationInterfaceRegistering` as part of this PR for completeness?

**Review by jonatack — COMMENTED** [review]
Light ACK 8eed7dd722177c2e36645d66e2c8254f9037a0e3

Modulo a question below.

**Comment by ryanofsky** [comment]
I don't understand what this is doing in the context of https://github.com/bitcoin-core/gui/pull/444. It looks like the `NetWatchLogModel::~NetWatchLogModel` destructor there calls `UnregisterValidationInterface`, which (because of the change in this PR) calls `NetWatchValidationInterface::ValidationInterfaceUnregistering` which calls `NetWatchLogModel::OrphanedValidationInterface` which then does `delete m_validation_interface` before the `~NetWatchLogModel` repeats `delete m_validation_interface` again which is a no-op.

But maybe the point of this is not actually to hook the `UnregisterValidationInterface` call but the `UnregisterAllValidationInterfaces` call? I'm not sure why that would be necessary either though. If the `UnregisterAllValidationInterfaces` call is happening before the `~NetWatchLogModel` destructor, it seems harmless but pointless to delete the validationinterface early. If destructor runs after `UnregisterAll`, the destructor is already unregistering immediately so there would no `UnregisterAll` callback would happen at all.

So I'm confused what new callback is doing and think PR could benefit from a description.

Probably also in #444 GUI code shouldn't be making low level-validationinterface calls. Ideally it would use something like the [`interfaces::Node::handleNotifyBlockTip`](https://github.com/bitcoin/bitcoin/blob/73b61717a977fc9d23f1bae3f8620641a9dee1f3/src/interfaces/node.h#L256-L259) function or [`interfaces::Chain::Notifications::updatedBlockTip` and `transactionAddedToMempool` functions](https://github.com/bitcoin/bitcoin/blob/73b61717a977fc9d23f1bae3f8620641a9dee1f3/src/interfaces/chain.h#L258-L269) but that is probably tangential to whatever issue this PR is solving.

**Comment by fanquake** [comment]
@luke-jr are you planning on either adding to the PR description to better explain the changes, or answering the review questions left here?

**Comment by fanquake** [comment]
No response to review questions. Still no PR description. Can be re-opened if that is going to change.

**Pull request closed without merging**